masonry_winit/src/vello_util.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2025 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Simple helpers for managing wgpu state and surfaces.
+
+use masonry_core::vello::{Error, wgpu};
+use wgpu::{
+    BlendComponent, BlendFactor, BlendState, CompositeAlphaMode, Device, Instance, PresentMode,
+    Surface, SurfaceConfiguration, Texture, TextureFormat, TextureUsages, TextureView,
+    util::{TextureBlitter, TextureBlitterBuilder},
+};
+
+/// Simple render context that maintains wgpu state for rendering the pipeline.
+pub(crate) struct RenderContext {
+    pub instance: Instance,
+    pub devices: Vec<DeviceHandle>,
+}
+
+pub(crate) struct DeviceHandle {
+    adapter: wgpu::Adapter,
+    pub device: Device,
+    pub queue: wgpu::Queue,
+}
+
+impl RenderContext {
+    pub(crate) fn new() -> Self {
+        let backends = wgpu::Backends::from_env().unwrap_or_default();
+        let flags = wgpu::InstanceFlags::from_build_config().with_env();
+        let backend_options = wgpu::BackendOptions::from_env_or_default();
+        let instance = Instance::new(&wgpu::InstanceDescriptor {
+            backends,
+            flags,
+            backend_options,
+        });
+        Self {
+            instance,
+            devices: Vec::new(),
+        }
+    }
+
+    /// Creates a new surface for the specified window and dimensions.
+    pub(crate) async fn create_surface<'w>(
+        &mut self,
+        window: impl Into<wgpu::SurfaceTarget<'w>>,
+        width: u32,
+        height: u32,
+        present_mode: PresentMode,
+    ) -> Result<RenderSurface<'w>, Error> {
+        self.create_render_surface(
+            self.instance.create_surface(window.into())?,
+            width,
+            height,
+            present_mode,
+        )
+        .await
+    }
+
+    /// Creates a new render surface for the specified window and dimensions.
+    pub(crate) async fn create_render_surface<'w>(
+        &mut self,
+        surface: Surface<'w>,
+        width: u32,
+        height: u32,
+        present_mode: PresentMode,
+    ) -> Result<RenderSurface<'w>, Error> {
+        let dev_id = self
+            .device(Some(&surface))
+            .await
+            .ok_or(Error::NoCompatibleDevice)?;
+
+        let device_handle = &self.devices[dev_id];
+        let capabilities = surface.get_capabilities(&device_handle.adapter);
+        let format = capabilities
+            .formats
+            .into_iter()
+            .find(|it| matches!(it, TextureFormat::Rgba8Unorm | TextureFormat::Bgra8Unorm))
+            .ok_or(Error::UnsupportedSurfaceFormat)?;
+
+        const PREMUL_BELND_STATE: BlendState = BlendState {
+            alpha: BlendComponent::REPLACE,
+            color: BlendComponent {
+                src_factor: BlendFactor::SrcAlpha,
+                dst_factor: BlendFactor::Zero,
+                operation: wgpu::BlendOperation::Add,
+            },
+        };
+        let (alpha_mode, blitter) = if capabilities
```

Comment:
I'd also like a warning for a transparent base colour without the `with_transparent(true)` window attribute set. This can be a TODO comment.